sksurv/metrics.py

```diff
@@ -411,65 +421,109 @@ def cumulative_dynamic_auc(survival_train, survival_test, estimate, times, tied_
+        estimate=numpy.tile(numpy.expand_dims(estimate, axis=1), (1, n_times, ))
+    
+    if times.max() >= test_time.max() or times.min() < test_time.min():
+        raise ValueError(
+            'all times must be within follow-up time of test data: [{}; {}['.format(
+                test_time.min(), test_time.max()))
 
+#   fit and transform IPCW
     cens = CensoringDistributionEstimator()
     cens.fit(survival_train)
     ipcw = cens.predict_ipcw(survival_test)
 
+#   expand arrays to (n_samples, n_times, ) shape      
     n_samples = test_time.shape[0]
-    scores = numpy.empty(times.shape[0], dtype=float)
-    for k, t in enumerate(times):
-        is_case = (test_time <= t) & test_event
-        is_control = test_time > t
-        n_controls = is_control.sum()
-
-        true_pos = []
-        false_pos = []
-        tp_value = 0.0
-        fp_value = 0.0
-        est_prev = numpy.infty
-
-        for i in range(n_samples):
-            est = estimate[i]
-            if numpy.absolute(est - est_prev) > tied_tol:
-                true_pos.append(tp_value)
-                false_pos.append(fp_value)
-                est_prev = est
-            if is_case[i]:
-                tp_value += ipcw[i]
-            elif is_control[i]:
-                fp_value += 1
-        true_pos.append(tp_value)
-        false_pos.append(fp_value)
-
-        sens = numpy.array(true_pos) / ipcw[is_case].sum()
-        fpr = numpy.array(false_pos) / n_controls
-        scores[k] = trapz(sens, fpr)
+    test_time = numpy.tile(numpy.expand_dims(test_time, axis=1), (1, n_times, ))
+    test_event = numpy.tile(numpy.expand_dims(test_event, axis=1), (1, n_times, ))
+    times = numpy.tile(numpy.expand_dims(times, axis=0), (n_samples, 1))
+    survival_test = numpy.tile(numpy.expand_dims(survival_test, axis=1), (1, n_times, ))
+    ipcw = numpy.tile(numpy.expand_dims(ipcw, axis=1), (1, n_times, ))
+
+#   check estimate after expanding test_time
+    estimate = _check_estimate2D(estimate, test_time)
+    
+#   sort estimates in ascending order and the other arrays too
+    o = numpy.argsort(-estimate, axis=0, )
+    test_time = numpy.take_along_axis(test_time, o, axis=0, )
+    test_event = numpy.take_along_axis(test_event, o, axis=0, )
+    estimate = numpy.take_along_axis(estimate, o, axis=0, )
+    survival_test = numpy.take_along_axis(survival_test, o, axis=0, )
+    ipcw = numpy.take_along_axis(ipcw, o, axis=0, )
+
+    is_case = numpy.logical_and(numpy.less_equal(test_time, times, ), 
+                            test_event)
+    is_control = numpy.greater_equal(test_time, times)
+    n_controls = is_control.sum(axis=0, )
+    is_tied = numpy.less_equal(
+                    numpy.absolute(numpy.subtract
+                                     (
+                                             estimate, numpy.roll(estimate, 1, axis=0, )
+                                     )
+                                 )
+                             , tied_tol)
```

Comment:
I don't think this construct works all of the time, because `numpy.roll(estimate, 1, axis=0)` moves that last element of `estimate` into the first position, thus the first element of `is_tied` could be True, whereas this never could be the case with the previous code. A unit test would definitely be required here.